port/linux/omrvmem.c

```diff
@@ -961,21 +966,54 @@ default_pageSize_reserve_memory(struct OMRPortLibrary *portLibrary, void *addres
 	 * Any changes made here may need to be reflected in that version.
 	 */
 	int fd = -1;
-	int flags = MAP_PRIVATE;
+	int ft = -1;
+	int flags = 0;
 	void *result = NULL;
 	int protectionFlags = PROT_NONE;
+	BOOLEAN useBackingSharedFile = FALSE;
+	BOOLEAN useBackingFile = FALSE;
 
 	Trc_PRT_vmem_default_reserve_entry(address, byteAmount);
 
+	if(mode & OMRPORT_VMEM_MEMORY_MODE_SHARE_FILE_OPEN) {
+		flags |= MAP_SHARED;
+		useBackingSharedFile = TRUE;
+	} else {
 #if defined(MAP_ANONYMOUS)
-	flags |= MAP_ANONYMOUS;
+		flags |= MAP_ANONYMOUS | MAP_PRIVATE;
+		useBackingSharedFile = FALSE;
 #elif defined(MAP_ANON)
-	flags |= MAP_ANON;
+		flags |= MAP_ANON | MAP_PRIVATE;
+		useBackingSharedFile = FALSE;
 #else
-	fd = portLibrary->file_open(portLibrary, "/dev/zero", EsOpenRead | EsOpenWrite, 0);
-	if (-1 != fd)
+		useBackingFile = TRUE;
+		flags |= MAP_PRIVATE;
 #endif
-	{
+	}
+
+	if (useBackingSharedFile) {
+		char filename[FILE_NAME_SIZE + 1];
+		sprintf(filename, "omrvmem_temp_%zu_%09d_%zu",  omrtime_current_time_millis(portLibrary), 
+								getpid(), 
+								pthread_self()); 
+		fd = shm_open(filename, O_RDWR | O_CREAT, 0600);
+		identifier->fd = fd;
+		shm_unlink(filename);
+		ft = ftruncate(fd, byteAmount);
```

Comment:
Is it safe to call `ftruncate` if `fd` == -1?

Reply:
The documentation does not say much, but I tested and when we call `ftruncate` and `fd == -1` the call returns -1 which is a fail and sets `errno` to 9 which is `EBADF` (Bad file number)